From 3cce57c19db4888884a5507b3e1d2535fe983b39 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 15:22:05 -0400
Subject: [infiniband] ocrdma: Allow only SEND opcode in case of UD QPs

Message-id: <d446f2695852b6731c2e7d6276fcef112ceb7d91.1411571481.git.dledford@redhat.com>
Patchwork-id: 96653
O-Subject: [Patch RHEL7 30/48] RDMA/ocrdma: Allow only SEND opcode in case of UD QPs
Bugzilla: 1081067 1088577
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Mitesh Ahuja <mitesh.ahuja@emulex.Com>

Prevent posting opcodes other than send and send immediate on the UD QPs.

Signed-off-by: Mitesh Ahuja <mitesh.ahuja@emulex.Com>
Signed-off-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Selvin Xavier <selvin.xavier@emulex.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit f252b5dc36e26368c7161f32ef304c30cd2d1f6c)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/ocrdma/ocrdma_verbs.c b/drivers/infiniband/hw/ocrdma/ocrdma_verbs.c
index 2b68235..7f54d24 100644
--- a/drivers/infiniband/hw/ocrdma/ocrdma_verbs.c
+++ b/drivers/infiniband/hw/ocrdma/ocrdma_verbs.c
@@ -2055,6 +2055,13 @@ int ocrdma_post_send(struct ib_qp *ibqp, struct ib_send_wr *wr,
  }
 
  while (wr) {
+  if (qp->qp_type == IB_QPT_UD &&
+      (wr->opcode != IB_WR_SEND &&
+       wr->opcode != IB_WR_SEND_WITH_IMM)) {
+   *bad_wr = wr;
+   status = -EINVAL;
+   break;
+  }
   if (ocrdma_hwq_free_cnt(&qp->sq) == 0 ||
       wr->num_sge > qp->sq.max_sges) {
    *bad_wr = wr;
-- 
1.7.1